File: tests/UniformSingleLayer2d.py
import unittest

import FrictionQPotFEM
import GMatElastoPlasticQPot.Cartesian2d as GMat
import GooseFEM
import numpy as np


class test_Generic2d(unittest.TestCase):
    """
    Tests
    """

    def test_eventDrivenSimpleShear(self):

        mesh = GooseFEM.Mesh.Quad4.Regular(3, 3)
        nelem = mesh.nelem()
        coor = mesh.coor()
        dofs = mesh.dofs()
        dofs[mesh.nodesLeftOpenEdge(), ...] = dofs[mesh.nodesRightOpenEdge(), ...]

        system = FrictionQPotFEM.UniformSingleLayer2d.System(
            coor,
            mesh.conn(),
            dofs,
            dofs[np.concatenate((mesh.nodesBottomEdge(), mesh.nodesTopEdge())), :].ravel(),
            [0, 1, 2, 6, 7, 8],
            [3, 4, 5],
        )

        nelas = system.elastic().size
        nplas = system.plastic().size

        epsy = 1e-3 * np.cumsum(np.ones((nplas, 5)), axis=1)

        system.setMassMatrix(np.ones(nelem))
        system.setDampingMatrix(np.ones(nelem))
        system.setElastic(np.ones(nelas), np.ones(nelas))
        system.setPlastic(np.ones(nplas), np.ones(nplas), epsy)
        system.setDt(1.0)

        for loop in range(2):

            if loop == 0:
                system.initEventDrivenSimpleShear()
                delta_u = system.eventDriven_deltaU()
            else:
                system.eventDriven_setDeltaU(delta_u)
                system.setU(np.zeros_like(coor))

            directions = [1, 1, 1, 1, 1, 1, -1, -1, -1]
            kicks = [False, False, True, False, True, False, False, True, False]
            indices = [0, 0, 0, 1, 1, 2, 1, 1, 0]
            multi = [-1, -1, +1, -1, +1, -1, +1, -1, +1]

            for direction, kick, index, m in zip(directions, kicks, indices, multi):

                eps_expect = epsy[0, index] + m * 0.5 * 1e-4
                system.eventDrivenStep(1e-4, kick, direction)
                self.assertTrue(np.allclose(GMat.Epsd(system.plastic_Eps()), eps_expect))
                self.assertTrue(system.residual() < 1e-5)

    def test_eventDrivenSimpleShear_historic(self):

        mesh = GooseFEM.Mesh.Quad4.Regular(3, 3)
        nelem = mesh.nelem()
        coor = mesh.coor()
        dofs = mesh.dofs()
        dofs[mesh.nodesLeftOpenEdge(), ...] = dofs[mesh.nodesRightOpenEdge(), ...]

        system = FrictionQPotFEM.UniformSingleLayer2d.System(
            coor,
            mesh.conn(),
            dofs,
            dofs[np.concatenate((mesh.nodesBottomEdge(), mesh.nodesTopEdge())), :].ravel(),
            [0, 1, 2, 6, 7, 8],
            [3, 4, 5],
        )

        nelas = system.elastic().size
        nplas = system.plastic().size

        epsy = 1e-1 * np.cumsum(np.random.random((nplas, 2000)), axis=1)

        system.setMassMatrix(np.ones(nelem))
        system.setDampingMatrix(np.ones(nelem))
        system.setElastic(np.ones(nelas), np.ones(nelas))
        system.setPlastic(np.ones(nplas), np.ones(nplas), epsy)
        system.setDt(1.0)

        system.initEventDrivenSimpleShear()

        kicks = np.zeros(1800, dtype=bool)
        kicks[1::2] = True

        for kick in kicks:

            u = system.u()
            epsy = system.plastic_CurrentYieldRight()
            eps_n = GMat.Epsd(system.plastic_Eps())
            idx_n = system.plastic_CurrentIndex()

            system.eventDrivenStep(1e-4, kick)
            idx = system.plastic_CurrentIndex()
            eps = GMat.Epsd(system.plastic_Eps())
            delta_u = system.u() - u
            system.setU(u)

            # note that this implementation is flawed in negative direction
            system.addSimpleShearEventDriven(1e-4, kick)
            check_idx = system.plastic_CurrentIndex()
            check_eps = GMat.Epsd(system.plastic_Eps())
            check_u = system.u() - u

            if not np.allclose(delta_u, check_u):
                self.assertTrue(np.all(idx == check_idx))
                i = np.argmin(epsy - eps_n)
                if kick:
                    self.assertTrue(np.sum(idx.astype(int) - idx_n.astype(int)) <= 4)
                    self.assertTrue(np.sum(check_idx.astype(int) - idx_n.astype(int)) <= 4)
                    self.assertTrue(np.isclose(eps.ravel()[i], epsy.ravel()[i] + 0.5 * 1e-4))
                    # note: only failing here, by about 1e-5 (so small error)
                    self.assertTrue(np.isclose(check_eps.ravel()[i], epsy.ravel()[i] + 0.5 * 1e-4))
                else:
                    self.assertTrue(np.sum(idx.astype(int) - idx_n.astype(int)) == 0)
                    self.assertTrue(np.sum(check_idx.astype(int) - idx_n.astype(int)) == 0)
                    self.assertTrue(np.isclose(eps.ravel()[i], epsy.ravel()[i] - 0.5 * 1e-4))
                    self.assertTrue(np.isclose(check_eps.ravel()[i], epsy.ravel()[i] - 0.5 * 1e-4))

            else:
                self.assertTrue(np.allclose(delta_u, check_u))


if __name__ == "__main__":

    unittest.main()

File: tests/UniformSingleLayer2d.cpp
#include <FrictionQPotFEM/UniformSingleLayer2d.h>
#include <catch2/catch_all.hpp>
#include <iostream>
#include <xtensor/xrandom.hpp>

TEST_CASE("FrictionQPotFEM::UniformSingleLayer2d", "UniformSingleLayer2d.h")
{
    SECTION("version")
    {
        std::cout << FrictionQPotFEM::version() << std::endl;

        auto deps = FrictionQPotFEM::UniformSingleLayer2d::version_dependencies();

        for (auto& i : deps) {
            std::cout << i << std::endl;
        }
    }

    SECTION("System::addAffineSimpleShear")
    {
        GooseFEM::Mesh::Quad4::Regular mesh(3, 3);

        xt::xtensor<size_t, 1> elas = {0, 1, 2, 6, 7, 8};
        xt::xtensor<size_t, 1> plas = {3, 4, 5};
        xt::xtensor<double, 2> epsy = {
            {1.0, 2.0, 3.0, 4.0}, {1.0, 2.0, 3.0, 4.0}, {1.0, 2.0, 3.0, 4.0}};

        FrictionQPotFEM::UniformSingleLayer2d::System sys(
            mesh.coor(),
            mesh.conn(),
            mesh.dofs(),
            xt::eval(xt::arange<size_t>(mesh.nnode() * mesh.ndim())),
            elas,
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({elas.size()}))),
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({elas.size()}))),
            plas,
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({plas.size()}))),
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({plas.size()}))),
            FrictionQPotFEM::epsy_initelastic_toquad(epsy),
            1,
            1,
            1,
            0);

        for (size_t i = 0; i < 10; ++i) {
            double delta_gamma = 0.01;
            double gamma = delta_gamma * static_cast<double>(i + 1);
            auto du = sys.affineSimpleShear(delta_gamma);
            sys.setU(xt::eval(sys.u() + du));
            REQUIRE(xt::allclose(xt::view(sys.Eps(), xt::all(), xt::all(), 0, 1), gamma));
        }
    }

    SECTION("System::addAffineSimpleShearCentered")
    {
        GooseFEM::Mesh::Quad4::Regular mesh(3, 3);

        xt::xtensor<size_t, 1> elas = {0, 1, 2, 6, 7, 8};
        xt::xtensor<size_t, 1> plas = {3, 4, 5};
        xt::xtensor<double, 2> epsy = {
            {1.0, 2.0, 3.0, 4.0}, {1.0, 2.0, 3.0, 4.0}, {1.0, 2.0, 3.0, 4.0}};

        FrictionQPotFEM::UniformSingleLayer2d::System sys(
            mesh.coor(),
            mesh.conn(),
            mesh.dofs(),
            xt::eval(xt::arange<size_t>(mesh.nnode() * mesh.ndim())),
            elas,
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({elas.size()}))),
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({elas.size()}))),
            plas,
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({plas.size()}))),
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({plas.size()}))),
            FrictionQPotFEM::epsy_initelastic_toquad(epsy),
            1,
            1,
            1,
            0);

        auto plastic = sys.plastic();
        auto conn = sys.conn();
        auto bot = xt::view(conn, xt::keep(plastic), 0); // missing last node, but ok for test
        auto top = xt::view(conn, xt::keep(plastic), 3);

        for (size_t i = 0; i < 10; ++i) {
            double delta_gamma = 0.01;
            double gamma = delta_gamma * static_cast<double>(i + 1);
            auto dus = sys.affineSimpleShearCentered(delta_gamma);
            sys.setU(xt::eval(sys.u() + dus));
            auto u = sys.u();
            auto du = xt::eval(xt::view(u, xt::keep(top), 1) + xt::view(u, xt::keep(bot), 1));
            REQUIRE(xt::allclose(xt::view(sys.Eps(), xt::all(), xt::all(), 0, 1), gamma));
            REQUIRE(xt::allclose(du, 0.0));
        }
    }

    SECTION("System::addSimpleShearEventDriven")
    {
        GooseFEM::Mesh::Quad4::Regular mesh(1, 1);

        xt::xtensor<size_t, 1> elas = {};
        xt::xtensor<size_t, 1> plas = {0};
        xt::xtensor<double, 2> epsy = {{1.0, 2.0, 3.0, 4.0}};

        FrictionQPotFEM::UniformSingleLayer2d::System sys(
            mesh.coor(),
            mesh.conn(),
            mesh.dofs(),
            xt::eval(xt::arange<size_t>(mesh.nnode() * mesh.ndim())),
            elas,
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({elas.size()}))),
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({elas.size()}))),
            plas,
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({plas.size()}))),
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({plas.size()}))),
            FrictionQPotFEM::epsy_initelastic_toquad(epsy),
            1,
            1,
            1,
            0);

        REQUIRE(sys.isHomogeneousElastic());

        double delta_eps = 1e-3;
        auto dV = sys.quad().AsTensor<2>(sys.dV());
        xt::xtensor<double, 2> Epsbar = xt::average(sys.Eps(), dV, {0, 1});
        REQUIRE(xt::allclose(GMatElastoPlasticQPot::Cartesian2d::Epsd(Epsbar)(), 0.0));

        sys.initEventDrivenSimpleShear();

        sys.eventDrivenStep(delta_eps, false, +1.0);

        Epsbar = xt::average(sys.Eps(), dV, {0, 1});
        REQUIRE(xt::allclose(
            GMatElastoPlasticQPot::Cartesian2d::Epsd(Epsbar)(), 1.0 - delta_eps / 2.0));

        sys.eventDrivenStep(delta_eps, true, +1.0);

        Epsbar = xt::average(sys.Eps(), dV, {0, 1});
        REQUIRE(xt::allclose(
            GMatElastoPlasticQPot::Cartesian2d::Epsd(Epsbar)(), 1.0 + delta_eps / 2.0));

        sys.eventDrivenStep(delta_eps, false, +1.0);

        Epsbar = xt::average(sys.Eps(), dV, {0, 1});
        REQUIRE(xt::allclose(
            GMatElastoPlasticQPot::Cartesian2d::Epsd(Epsbar)(), 2.0 - delta_eps / 2.0));

        sys.eventDrivenStep(delta_eps, true, +1.0);

        Epsbar = xt::average(sys.Eps(), dV, {0, 1});
        REQUIRE(xt::allclose(
            GMatElastoPlasticQPot::Cartesian2d::Epsd(Epsbar)(), 2.0 + delta_eps / 2.0));

        sys.eventDrivenStep(delta_eps, false, -1.0);

        Epsbar = xt::average(sys.Eps(), dV, {0, 1});
        REQUIRE(xt::allclose(
            GMatElastoPlasticQPot::Cartesian2d::Epsd(Epsbar)(), 2.0 + delta_eps / 2.0));

        sys.eventDrivenStep(delta_eps, true, -1.0);

        Epsbar = xt::average(sys.Eps(), dV, {0, 1});
        REQUIRE(xt::allclose(
            GMatElastoPlasticQPot::Cartesian2d::Epsd(Epsbar)(), 2.0 - delta_eps / 2.0));

        sys.eventDrivenStep(delta_eps, false, -1.0);

        Epsbar = xt::average(sys.Eps(), dV, {0, 1});
        REQUIRE(xt::allclose(
            GMatElastoPlasticQPot::Cartesian2d::Epsd(Epsbar)(), 1.0 + delta_eps / 2.0));

        sys.eventDrivenStep(delta_eps, true, -1.0);

        Epsbar = xt::average(sys.Eps(), dV, {0, 1});
        REQUIRE(xt::allclose(
            GMatElastoPlasticQPot::Cartesian2d::Epsd(Epsbar)(), 1.0 - delta_eps / 2.0));
    }

    SECTION("System::addSimpleShearToFixedStress")
    {
        GooseFEM::Mesh::Quad4::Regular mesh(1, 1);

        xt::xtensor<size_t, 1> elas = {};
        xt::xtensor<size_t, 1> plas = {0};
        xt::xtensor<double, 2> epsy = {{1.0, 2.0, 3.0, 4.0}};

        FrictionQPotFEM::UniformSingleLayer2d::System sys(
            mesh.coor(),
            mesh.conn(),
            mesh.dofs(),
            xt::eval(xt::arange<size_t>(mesh.nnode() * mesh.ndim())),
            elas,
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({elas.size()}))),
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({elas.size()}))),
            plas,
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({plas.size()}))),
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({plas.size()}))),
            FrictionQPotFEM::epsy_initelastic_toquad(epsy),
            1,
            1,
            1,
            0);

        REQUIRE(sys.isHomogeneousElastic());

        auto dV = sys.quad().AsTensor<2>(sys.dV());
        xt::xtensor<double, 2> Epsbar = xt::average(sys.Eps(), dV, {0, 1});
        xt::xtensor<double, 2> Sigbar = xt::average(sys.Sig(), dV, {0, 1});
        REQUIRE(xt::allclose(GMatElastoPlasticQPot::Cartesian2d::Epsd(Epsbar)(), 0.0));
        REQUIRE(xt::allclose(GMatElastoPlasticQPot::Cartesian2d::Sigd(Sigbar)(), 0.0));

        double target_stress = 1.0;
        sys.addSimpleShearToFixedStress(target_stress);
        Epsbar = xt::average(sys.Eps(), dV, {0, 1});
        Sigbar = xt::average(sys.Sig(), dV, {0, 1});
        REQUIRE(
            xt::allclose(GMatElastoPlasticQPot::Cartesian2d::Epsd(Epsbar)(), target_stress / 2.0));
        REQUIRE(xt::allclose(GMatElastoPlasticQPot::Cartesian2d::Sigd(Sigbar)(), target_stress));

        target_stress = 1.5;
        sys.addSimpleShearToFixedStress(target_stress);
        Epsbar = xt::average(sys.Eps(), dV, {0, 1});
        Sigbar = xt::average(sys.Sig(), dV, {0, 1});
        REQUIRE(
            xt::allclose(GMatElastoPlasticQPot::Cartesian2d::Epsd(Epsbar)(), target_stress / 2.0));
        REQUIRE(xt::allclose(GMatElastoPlasticQPot::Cartesian2d::Sigd(Sigbar)(), target_stress));

        target_stress = 0.5;
        sys.addSimpleShearToFixedStress(target_stress);
        Epsbar = xt::average(sys.Eps(), dV, {0, 1});
        Sigbar = xt::average(sys.Sig(), dV, {0, 1});
        REQUIRE(
            xt::allclose(GMatElastoPlasticQPot::Cartesian2d::Epsd(Epsbar)(), target_stress / 2.0));
        REQUIRE(xt::allclose(GMatElastoPlasticQPot::Cartesian2d::Sigd(Sigbar)(), target_stress));
    }

    SECTION("System::triggerElementWithLocalSimpleShear")
    {
        GooseFEM::Mesh::Quad4::Regular mesh(3, 3);

        auto dofs = mesh.dofs();
        auto top = mesh.nodesTopEdge();
        auto bottom = mesh.nodesBottomEdge();
        size_t nfix = top.size();
        xt::xtensor<size_t, 1> iip = xt::empty<size_t>({2 * mesh.ndim() * nfix});
        xt::view(iip, xt::range(0 * nfix, 1 * nfix)) = xt::view(dofs, xt::keep(bottom), 0);
        xt::view(iip, xt::range(1 * nfix, 2 * nfix)) = xt::view(dofs, xt::keep(bottom), 1);
        xt::view(iip, xt::range(2 * nfix, 3 * nfix)) = xt::view(dofs, xt::keep(top), 0);
        xt::view(iip, xt::range(3 * nfix, 4 * nfix)) = xt::view(dofs, xt::keep(top), 1);

        xt::xtensor<size_t, 1> elas = {0, 1, 2, 3, 5, 6, 7, 8};
        xt::xtensor<size_t, 1> plas = {4};
        xt::xtensor<double, 2> epsy = {{1.0, 2.0, 3.0, 4.0}};

        FrictionQPotFEM::UniformSingleLayer2d::System sys(
            mesh.coor(),
            mesh.conn(),
            dofs,
            iip,
            elas,
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({elas.size()}))),
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({elas.size()}))),
            plas,
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({plas.size()}))),
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({plas.size()}))),
            FrictionQPotFEM::epsy_initelastic_toquad(epsy),
            1,
            1,
            1,
            0);

        REQUIRE(sys.isHomogeneousElastic());

        double delta_eps = 1e-3;
        sys.triggerElementWithLocalSimpleShear(delta_eps, 0);

        auto eps = GMatElastoPlasticQPot::Cartesian2d::Epsd(sys.Eps());
        auto plastic = sys.plastic();
        auto eps_p = xt::view(eps, xt::keep(plastic), xt::all());

        REQUIRE(xt::allclose(eps_p, 1.0 + delta_eps / 2.0));
    }
}
